actors/builtin/market/market_actor.go

```diff
@@ -111,7 +113,10 @@ func (a Actor) WithdrawBalance(rt Runtime, params *WithdrawBalanceParams) *adt.E
 func (a Actor) AddBalance(rt Runtime, providerOrClientAddress *addr.Address) *adt.EmptyValue {
 	msgValue := rt.Message().ValueReceived()
 
-	nominal, _ := escrowAddress(rt, *providerOrClientAddress)
+	nominal, _, codeId, approved := escrowAddress(rt, *providerOrClientAddress)
+	if codeId.Equals(builtin.StorageMinerActorCodeID) {
+		rt.ValidateImmediateCallerIs(approved...)
+	}
```

Comment:
There must be an `else` here so that caller validation is always performed. [Edit] I see you left it behind in `escrowAddress` but it's too confusing to have validation in different places with different parameters.

It doesn't make sense that for clients we allow any party to deposit funds, but for providers we require it to be the worker/owner. So for AddBalance, just always use `ValidateImmediateCallerType(builtin.CallerTypesSignable...)`

There will then be no need to return the code id from `escrowAddress`.

Reply:
This is done.